[state_dict][4/N] Support strict flag for model.load_state_dict

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111111
* #111120
* #111110
* __->__ #111109
* #111275
* #111107
* #111106

As title

Differential Revision: [D50209723](https://our.internmc.facebook.com/intern/diff/D50209723/)

fegin committed:
[state_dict][4/N] Support strict flag for model.load_state_dict

As title

Differential Revision: [D50209723](https://our.internmc.facebook.com/intern/diff/D50209723/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][4/N] Support strict flag for model.load_state_dict"


As title

Differential Revision: [D50209723](https://our.internmc.facebook.com/intern/diff/D50209723/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][4/N] Support strict flag for model.load_state_dict"


As title

Differential Revision: [D50209723](https://our.internmc.facebook.com/intern/diff/D50209723/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/checkpoint/test_state_dict.py:
```diff
@@ -355,3 +355,18 @@ def init_model_optim():
+        model_state_dict, _ = state_dict(model)
+        key = next(iter(model_state_dict.keys()))
+        model_state_dict.pop(key)
+        with self.assertRaisesRegex(RuntimeError, "Missing key"):
```
might also wanna test unexpected keys, i.e. an extra key in the state dict to load?

rohan-varma commented on test/distributed/checkpoint/test_state_dict.py:
```diff
@@ -355,3 +355,18 @@ def init_model_optim():
+        model_state_dict.pop(key)
+        with self.assertRaisesRegex(RuntimeError, "Missing key"):
+            load_state_dict(model, model_state_dict=model_state_dict)
+        load_state_dict(
```
I believe the nn.module.load_state_dict will return `missing_keys, unexpected_keys` tuple, will this API do the same?

fegin replied:
No, it will not. We can add the support.

fegin replied:
I'll add the return values in a separate PR.

fegin committed:
Update on "[state_dict][4/N] Support strict flag for model.load_state_dict"


As title

Differential Revision: [D50209723](https://our.internmc.facebook.com/intern/diff/D50209723/)

[ghstack-poisoned]

wz337 reviewed (approved):
Accepted in phabricator.

fegin committed:
Update on "[state_dict][4/N] Support strict flag for model.load_state_dict"


As title

Differential Revision: [D50209723](https://our.internmc.facebook.com/intern/diff/D50209723/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][4/N] Support strict flag for model.load_state_dict"


As title

Differential Revision: [D50209723](https://our.internmc.facebook.com/intern/diff/D50209723/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][4/N] Support strict flag for model.load_state_dict"


As title

Differential Revision: [D50209723](https://our.internmc.facebook.com/intern/diff/D50209723/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][4/N] Support strict flag for model.load_state_dict"


As title

Differential Revision: [D50209723](https://our.internmc.facebook.com/intern/diff/D50209723/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
